### PR TITLE
feat: add gitopsapi delete operation

### PR DESCRIPTION
**Pull request by Kaitou786**
### Description

Exposed a different endpoint starting with `/_delete` to let the server know that the request is for delete object

### Breaking Change

- [ ] Yes
- [x] No

### Testing Notes

Created a gitopsapi object and sent a post request with a object to the delete endpoint to see it gets deleted
GitOpsApi object used:
```yaml
apiVersion: git.flanksource.com/v1
kind: GitopsAPI
metadata:
  name: gitopsapi-sample
spec:
  gitRepository: https://github.com/Kaitou786/test-gitops-api.git
  kustomization: clusters/dev01/kustomization.yaml
  secretRef:
    name: git-operator-test
  tokenRef:
    name: token
  branch: random
  searchPath: "clusters/"
  pullRequest:
    body: "my test PR"
    title: "some awesome title"
```

The post data
```jsom
{
  "apiVersion": "acmp.corp/v1",
  "kind": "NamespaceRequest",
  "metadata": {
    "name": "tenant3",
    "annotations": {
      "a": "b"
    },
    "labels": {
      "c": "d"
    }
  },
  "spec": {
    "memory": 10,
    "cluster": "dev01"
  },
  "cluster": "dev02"
}
```

The Curl request:
```
 curl -X POST 'localhost:8888/_delete/default/gitopsapi-sample/mypassword' -d @data.json
 ```

See my testing PR: https://github.com/Kaitou786/test-gitops-api/pull/15
**How you can replicate my testing:**
_How can the reviewer validate this PR?_

### **Links**

_Issues links or other related resources_

### CODE REVIEW

File: controllers/gitopsapi_controller.go [hunk 1]
```diff
@@ -72,6 +72,11 @@ func serve(c echo.Context, r *GitopsAPIReconciler) error {
 	name := c.Param("name")
 	namespace := c.Param("namespace")
 	token := c.Param("token")
+	var deleteObj bool
+	if strings.HasPrefix(c.Path(), "/_delete") {
+		r.Log.Info("called the delete endpoint will delete the specified objs")
+		deleteObj = true
+	}
```
Comment (moshloop) [inline controllers/gitopsapi_controller.go:79]:
```suggestion
	deleteObj := strings.HasPrefix(c.Path(), "/_delete")
```

File: controllers/gitopsapi_controller.go [hunk 2]
```diff
@@ -127,7 +132,7 @@ func GetKustomizaton(fs billy.Filesystem, path string) (*types.Kustomization, er
 	return &kustomization, nil
 }
 
-func HandleGitopsAPI(ctx context.Context, logger logr.Logger, git connectors.Connector, api gitv1.GitopsAPI, contents io.Reader) (hash string, pr int, err error) {
+func HandleGitopsAPI(ctx context.Context, logger logr.Logger, git connectors.Connector, api gitv1.GitopsAPI, contents io.Reader, deleteObj bool) (hash string, pr int, err error) {
```
Comment (moshloop) [inline controllers/gitopsapi_controller.go:135]: I would split this function into 2, 1 for Create/Update and another for delete with some extracted utility methods

**Comment by moshloop** [comment]
Can we also add an e2e ?

**Comment by Kaitou786** [comment]
See the diff PRs created by the tests:
https://github.com/flanksource/git-operator-test/pull/343/
https://github.com/flanksource/git-operator-test/pull/344/
https://github.com/flanksource/git-operator-test/pull/345

**Comment by Kaitou786** [comment]
Refactored a bit with some more functions. Not sure but let me know about this approach... PRs with this approach:
https://github.com/flanksource/git-operator-test/pull/349
https://github.com/flanksource/git-operator-test/pull/350
https://github.com/flanksource/git-operator-test/pull/351

File: controllers/gitopsapi_controller.go [hunk 3]
```diff
@@ -127,7 +148,143 @@ func GetKustomizaton(fs billy.Filesystem, path string) (*types.Kustomization, er
...
+	}
+	relativePath := strings.Replace(contentPath, path.Dir(api.Spec.Kustomization)+"/", "", -1)
+	if err = copy(body, contentPath, fs, work); err != nil {
+		return
+	}
+	index := findElement(kustomization.Resources, relativePath)
+	if index == -1 {
+		kustomization.Resources = append(kustomization.Resources, relativePath)
+	}
+	existingKustomization, _ := yaml.Marshal(kustomization)
```
Comment (BrendanGalloway) [inline controllers/gitopsapi_controller.go:195]: you might want to catch and check this error?

Comment (Kaitou786) [reply controllers/gitopsapi_controller.go]: Added a new commit for that

**Comment by Kaitou786** [comment]
subset of #46

**Pull request closed without merging**